### DISCUSSION TITLE
Easier embedding of Rust / into Rust

### DISCUSSION THREAD

**Discussion by llogiq** [koto-lang/koto#65] [General]
Hi there. I'm experimenting with embedding koto in another crate, and I feel the protocol around ExternalVaues (as showcased in the example; to my knowledge there are no other docs) is quite cumbersome.

I have built a procedural macro that can insert Rust functions into a given prelude, however, I'd like to also extend it to derive support for my own types (perhaps in conjunction with the proc macro also setting up a kind of vtable). What do you think about simplifying this interface to an `ExternalValue` + a `Box<dyn fn(KeyValue) -> ExternalFunction>` that will do the vtable dispatch? This might both be faster and easier to code against.

**Comment by irh** [comment]
Hi @llogiq, glad to hear that you're trying out Koto! 

I'm sorry that the docs are missing here, I started a documentation effort on the `help` branch but I haven't touched the topic of integrating with Rust yet and I've had a busy few weeks, I'm looking forward to picking this up again.

I agree that making custom types is a bit cumbersome at the moment, especially without a guide.. My thinking is that I'd like to make sure the fundamentals are correct and the implementation straight-forward, and then some more ergonomic helpers can be added to make usage a bit friendlier. 

I'll try to clarify the current approach:
- An `ExternalValue` is just a placeholder for some arbitrary Rust value. The intention is that scripts don't interact with them directly,  but instead to make them useful they get placed in a `ValueMap` as a hidden value using `ExternalDataId`. 
- An interface is then built up around the external value in the `ValueMap`. Functions and any other value type can be added to the map, like with any other map in Koto. 
- There are a [couple](https://github.com/koto-lang/koto/blob/c411fbc40d9b734a81b62154c40a801c00062318/src/runtime/src/core/thread.rs#L63) of [other](https://github.com/koto-lang/koto/blob/c411fbc40d9b734a81b62154c40a801c00062318/src/runtime/src/core/io.rs#L10) examples in the core library that can be looked at. The `File` example is more extensive. You can see that I added a `file_fn` helper for reducing boilerplate when adding functions. 

If I understand you correctly you'd rather expect that the `ExternalValue` itself would be the value that the user interacts with (instead of a `ValueMap`), with behaviour on the value exposed via a `KeyValue -> Function` mapper? 

If I've got that right I think I would rather extend the type via an embedded `ValueMap` (just to enable all value types, not just functions). The difference in approaches would then be that you would have an `ExternalValue` with an embedded `ValueMap`, rather than a `ValueMap` with an embedded `ExternalValue`... 

Does that make sense? I can see some value in inverting the approach here, it might make the expected usage more explicit (although maybe utilities and docs will help there).

**Comment by llogiq** [comment]
Adding a `ValueMap` field in the `ExternalValue` variant makes a lot of sense; those are `Arc` and can easily be generated as a lazy static to be cloned for further use.

**Comment by irh** [comment]
Yes I can see an advantage there, it'd be cheaper to clone the interface map than having to rebuild it for each new value.

**Comment by llogiq** [comment]
I have started a WIP PR #64. Do you think this is the right direction?

**Comment by irh** [comment]
Yes I think this is roughly how I would expect it to look, I appreciate you diving in!

I'm still unsure if I want to proceed in this direction though. The main concern I have is that this will introduce a new object-like-thing to Koto, and I’d like to avoid that if possible.

I'm thinking about an alternative solution which would be to extend `ValueMap` so that it includes an optional `ExternalValue`, plus an optional ‘immutable data’ map, which would act as fallback when values aren't found in the main data, and which would allow for the fast cloning of a shared function table that you're looking for.

`ValueMap` would then look something like:

```rust
/// The Map value type used in Koto
#[derive(Clone, Debug, Default)]
pub struct ValueMap {
    data: Arc<RwLock<ValueMapContents>>,
    immutable_data: Option<Arc<RwLock<ValueHashMap>>>,
    external_data: Option<Arc<RwLock<dyn ExternalValue>>>,
};
```
…and `ExternalValue` and `ExternalDataId` could be removed as `Value` variants.

Something I like about this is that modules like could expose their features as immutable. It’s never really felt right that there’s nothing to stop a user from saying `number.pi = -1` (but users could still extend core modules with their own functionality if they like).

…and I haven’t figured out a few details of the expected behaviour of map operations over a mix of mutable and immutable entries (like how map operations should apply), but I’d need to start playing with some examples to get a feel for the right direction.

**Reply by llogiq** [reply]
The downside of this would be another layer of indirection (and a map) for each ExternalValue.

I might live with that, but I'd like to try this variant, too and see how large the overhead really is.

I think the only thing left in this PR is the method call protocol. Can you give me a pointer where in the code to change this?

**Reply by irh** [reply]
> The downside of this would be another layer of indirection (and a map) for each ExternalValue.

I think the only additional indirection here would be the `Option`? I would assume this won't be a bottleneck for you given everything else that's going on in the runtime? The mutable `data` map (which you wouldn't be using by default) would be created for each value, but if this is too much overhead then there would be ways to optimize this too.. 

I think the key thing for me is to make sure that the semantics of an external value type are clear to the user. So far this has been achieved by wrapping up external values in Maps, so I'm trying to see if that can be maintained before extending the number of concepts in the language. Types with overloaded operators are just Maps, modules are just Maps, etc..

Now, it could be that behind the scenes there are different flavours of Map (data-only, overloaded operators, immutable, externally defined value), and then conversions happen on demand based on usage. I went in this direction with Number (which switches between Int and Float) and I have in mind potential optimizations along these lines for List (e.g. a List of Numbers takes up less space than a List of Values). So I wouldn't be too worried about performance at this stage, I'd prefer to focus on getting the general design right before optimizing.

> I think the only thing left in this PR is the method call protocol. Can you give me a pointer where in the code to change this?

You'll need to extend `Vm::run_access()` so that the accessed function gets placed in the target register. After that the function should just be called like any other (I think.. I might be forgetting something, let me know if you run into trouble).

**Reply by llogiq** [reply]
I understand that you want to keep the language simple, and yes, allowing an enum of Single-key-value, map would make most of the overhead moot. There would still be some memory overhead, but I think that shouldn't affect us too badly. But that's just conjecture at this point, we'll know when the measurement's in.

Thanks, I'll look into `Vm::run_access()`.

**Reply by irh** [reply]
It occurred to me that having a non-optional `ValueMap` in every `ExternalValue` is overhead that should be avoidable (e.g. if you want to have a list of many external values it would be good to avoid creating a map for each one), so the map in `ExternalValue` should either be optional, or there's going to be a need for a split between something like `ExternalData` (which is just the data) vs. `ExternalValue` (which includes associated items).

I'm still mulling over the options here, I haven't decided if I'm comfortable with having an 'external vs internal' split for object-like-things. It seems to me that both variants will want roughly the same feature set in the long run (locked vs unlocked, meta functions, etc.) so my natural inclination is to keep them unified. I can certainly see the need for improving the current situation though. I'm busy today but hoping to give this topic some proper attention tomorrow.

**Reply by llogiq** [reply]
There is some prior art in the lua implementation; they have [userdata and lightuserdata](https://stackoverflow.com/questions/4329643/what-is-userdata-and-lightuserdata-in-lua).

**Comment by irh** [comment]
@llogiq I hope you don't mind but I figured the topic was worth moving over to a discussion while we continue to discuss the design of a solution.

I'm curious to hear about the project that you're integrating Koto into? Is there anything you can share about it?

**Reply by llogiq** [reply]
Sure! We are building a configuration DSL for [synth](https://github.com/openquery-io/synth) which at some point would also include expressions and functions, so I looked at a few lang impls and liked koto best so far.

**Reply by irh** [reply]
Nice, looks interesting, and great name 😄! Let me know if you run into any problems. You're welcome to start discussions here asking general questions about the language, and you can also contact me by email if you prefer.

**Comment by irh** [comment]
I think I'm getting towards a design that I'm happy with:

- Introduce a new `MetaKey` variant: `Named(String)`.
  - This will allow for `MetaMap` entries with arbitrary names.
- Introduce a new `Value` variant: `ExternalData`.
  - This will be the 'data-only' version of `ExternalValue`.
- Add an `Arc<RwLock<MetaMap>>` to `ExternalValue`.
  - User-facing functions (and other value types) can then be added to the external value's meta map using `MetaKey::Named`.

The main advantage of this approach is that external values would then also be able to define meta behaviour like overloaded operators.

To extend the idea:

- Koto scripts could add named items to the meta map with the syntax `@'name'`, e.g.
```koto
  foo = 
    @type: 'Foo'
    @'bar': || 'Hello from meta-function "bar"'

  foo.bar()
  # Hello from meta-function "bar" 
```
- A helper could be added to inspect accessible named items on a value (currently this can be done with `map.keys()`), something like `koto.available_items(value)`.

@llogiq How does this sound to you?

**Reply by irh** [reply]
> Add an `Arc<RwLock<MetaMap>>` to `ExternalValue`.

This could  probably be `Arc<MetaMap>`, I can't think of a good reason at the moment  to modify metamaps after the value is created.

**Reply by llogiq** [reply]
Yeah, so basically I need to amend my PR to change the ValueMap vtable to a MetaMap (once that has arbitrary string keys), and then it's at least the third part.

I think this plan makes a lot of sense, and hope I can help to bring it to fruition.

**Reply by irh** [reply]
Great, I think it makes sense for me to start a branch in the next days that implements the MetaMap changes (I think there'll need to be some hoop-jumping to access the entries efficiently with `&str` similar to how it works with `ValueKeyRef`), then you'll be welcome to push it forward from there.

**Reply by llogiq** [reply]
Perfect! Just give me a ping when I can rebase my PR on top of your work. Thank you!

**Reply by irh** [reply]
@llogiq I've just merged #69 which includes named meta entries and should work as a foundation for the `ExternalValue` changes.

> I think there'll need to be some hoop-jumping to access the entries efficiently with &str

Re-reading this reminded me that I left a `TODO` to sort this out, that I forgot about before merging. I'll follow up on that as soon as I can as it will affect performance, but otherwise it shouldn't be in your way.

**Reply by irh** [reply]
> I'll follow up on that as soon as I can

Done in #70.

**Comment by irh** [comment]
@llogiq I finally managed to get to implementing a new `ExternalValue` value type based on this discussion in #73. Thanks for all your input!
